lib/kube/proxy/single_cert_handler_test.go

```diff
@@ -0,0 +1,168 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package proxy
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+	"net/http"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+
+	"github.com/gravitational/teleport/api/types"
+	testingkubemock "github.com/gravitational/teleport/lib/kube/proxy/testing/kube_server"
+)
+
+// pathRoutedKubeClient uses the given rest.Config to build a Kubernetes client
+// using path-based routing derived from the provided Teleport and Kubernetes
+// cluster names.
+func pathRoutedKubeClient(t *testing.T, restConfig *rest.Config, teleportCluster, kubeCluster string) *kubernetes.Clientset {
+	t.Helper()
+
+	encTeleportCluster := base64.RawURLEncoding.EncodeToString([]byte(teleportCluster))
+	encKubeCluster := base64.RawURLEncoding.EncodeToString([]byte(kubeCluster))
+	restConfig.Host += fmt.Sprintf("/v1/teleport/%s/%s", encTeleportCluster, encKubeCluster)
+
+	client, err := kubernetes.NewForConfig(restConfig)
+	require.NoError(t, err)
+
+	return client
+}
+
+func TestSingleCertRouting(t *testing.T) {
+	kubeMockA, err := testingkubemock.NewKubeAPIMock()
+	require.NoError(t, err)
+	t.Cleanup(func() { kubeMockA.Close() })
+
+	kubeMockB, err := testingkubemock.NewKubeAPIMock(
+		// This endpoint returns a known mock error so we can determine from the
+		// response which cluster the request was routed to.
+		testingkubemock.WithGetPodError(
+			metav1.Status{
+				Status:  metav1.StatusFailure,
+				Message: "cluster b error",
+				Reason:  metav1.StatusReasonInternalError,
+				Code:    http.StatusInternalServerError,
+			},
+		),
+	)
+	require.NoError(t, err)
+	t.Cleanup(func() { kubeMockB.Close() })
+
+	defaultRoleSpec := RoleSpec{
+		Name:       roleName,
+		KubeUsers:  roleKubeUsers,
+		KubeGroups: roleKubeGroups,
+	}
+
+	const clusterName = "root.example.com"
+
+	tests := []struct {
+		name string
+
+		kubeClusterOverride string
+		roleSpec            RoleSpec
+		assert              func(t *testing.T, restConfig *rest.Config)
+	}{
+		{
+			name:     "successful path routing to multiple clusters",
+			roleSpec: defaultRoleSpec,
+			assert: func(t *testing.T, restConfig *rest.Config) {
+				clientB := pathRoutedKubeClient(t, restConfig, clusterName, "b")
+				_, err = clientB.CoreV1().Pods(metav1.NamespaceDefault).List(context.Background(), metav1.ListOptions{})
+				require.NoError(t, err)
+
+				clientA := pathRoutedKubeClient(t, restConfig, clusterName, "a")
+				_, err := clientA.CoreV1().Pods(metav1.NamespaceDefault).List(context.Background(), metav1.ListOptions{})
+				require.NoError(t, err)
+
+			},
```

Comment:
This test triggers `ensureRouteNotOverwritten()` errors in a way that implies some state (`Identity.KubernetesCluster`, at least) is shared between requests. I haven't been able to reproduce this behavior on an actual cluster, so either:
1. Something about the test environment/clients shares the identity object in a way not representative of reality
2. The identity _is_ sometimes cached between requests (just not in a way I can trigger), in which case I think this method for passing path parameters around by overwriting identity fields may not be viable.

Will need to spend more time debugging this. 😕

Reply:
I found two issues with the tests:
- `GenTestKubeClientTLSCert` already fills all the kubernetes cluster + route to cluster information
- `restConfig` was edited by appending the route to cluster into the existing endpoint. This means that the second request to cluster A will have `/v1/teleport/<telecluster/<kubeclterA//v1/teleport/<telecluster/<kubeclterB/` as the host. that's why it fails
check the fixes here https://github.com/gravitational/teleport/commit/af3a02c22d675777136adb1872111e2f6bce9a28

Reply:
Oh, that was definitely it, thanks for finding the issue! I was losing my mind a bit over that and it was just the client config getting reused 🤦 

`GenTestKubeClientTLSCert` accepts `""` already for the KubernetesCluster value, and I believe most identities have `RouteToCluster` configured, so technically I think that was working, if a bit ugly. In any case, I'll add some test cases with an explicitly empty `RouteToCluster`.

Reply:
I added some extra tests explicitly tweaking RouteToCluster with session MFA enabled/disabled, so I think I'm happy with the state of things now.